### ISSUE TITLE
`conda update conda` destroyed the conda installation on Windows

### ISSUE THREAD

**Issue by ukoethe** [conda/conda#2325]
This now happened for the second time. Here is the log:

```
C:\Users\ukoethe> conda --version
conda 4.0.0

C:\Users\ukoethe> conda update conda
Fetching package metadata: ....
.Solving package specifications: ...........

Package plan for installation in environment C:\conda-test:

The following NEW packages will be INSTALLED:

    vs2015_runtime: 14.00.23026.0-0

The following packages will be UPDATED:

    conda:          4.0.0-py35_0 --> 4.0.5-py35_0
    python:         3.5.1-2      --> 3.5.1-4

Proceed ([y]/n)?

Extracting packages ...
[      COMPLETE      ]|##################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|##################################################| 100%
Linking packages ...
An unexpected error has occurred, please consider sending the            |  33%
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.

Traceback (most recent call last):
  File "C:\conda-test\Scripts\conda-script.py", line 5, in <module>
  File "C:\conda-test\lib\site-packages\conda\cli\main.py", line 139, in main
  File "C:\conda-test\lib\site-packages\conda\cli\main.py", line 146, in args_func
  File "C:\conda-test\lib\site-packages\conda\cli\main_update.py", line 48, in execute
  File "C:\conda-test\lib\site-packages\conda\cli\install.py", line 424, in install
  File "C:\conda-test\lib\site-packages\conda\plan.py", line 539, in execute_actions
  File "C:\conda-test\lib\site-packages\conda\instructions.py", line 149, in execute_instructions
  File "C:\conda-test\lib\site-packages\conda\instructions.py", line 92, in LINK_CMD
  File "C:\conda-test\lib\site-packages\conda\instructions.py", line 88, in link
  File "C:\conda-test\lib\site-packages\conda\install.py", line 631, in link
  File "C:\conda-test\lib\os.py", line 241, in makedirs
PermissionError: [WinError 5] Access denied: 'C:\\conda-test\\DLLs'

C:\Users\ukoethe> conda --version
command "conda" not found
```

It seems that the update attempts to replace the DLL directory that is used by the python process which runs conda. However, Windows disallows deleting or renaming open files. Shouldn't conda at least be able to roll-back changes if the update fails halfway through?

**Comment by ukoethe** [comment]
Interestingly, when I called `conda update conda` in a fresh installation from `Miniconda3-latest-Windows-x86_64.exe`, it worked. Strange.

**Comment by rcurnow** [comment]
I've just had the same experience. updating from version 3.19.0
guess i'll try installing again and see what happens
i'm new to all this

**Comment by rcurnow** [comment]
so with the new install, no updates were required,

**Comment by ccordoba12** [comment]
> It seems that the update attempts to replace the DLL directory that is used by the python process which runs conda. However, Windows disallows deleting or renaming open files

conda doesn't replace files, it just moves them to a special directory that will be deleted next time you run conda and those file are not in use. That's what allows conda to update things like `python` or `requests` in the root environment.

I think the problem you're seeing is caused because you installed Anaconda/Miniconda for all users, and ran conda in a non-privileged terminal. Then conda can't do its job of moving files in use because it doesn't have the right permissions to do it.

So the solution would be to run conda as Administrator. But let's ping @msarahan to see if he has something else to say about this :-)

**Comment by msarahan** [comment]
@ccordoba12 That's worth a try, but I don't think it is the solution.  This smells like a race condition to me - something in the DLLs folder is in use when Conda tries move/replace it.  I really don't understand how conda replaces those files without requiring a reboot.  There is no safe way that I know of.  Windows provides a way of replacing files that are in use, but only after a reboot.

**Comment by ccordoba12** [comment]
>  I really don't understand how conda replaces those files without requiring a reboot. There is no safe way that I know of. Windows provides a way of replacing files that are in use, but only after a reboot.

AFAIK, conda **doesn't** replace files (please see my previous answer). It moves files in use to a special directory and tries to remove them later. That allows conda to unpack new versions of the same files in the place they were before.

**Comment by msarahan** [comment]
Sure, but you can't move files if they're in use, no matter whether you move them to the trash directly, or to some intermediate.

**Comment by ccordoba12** [comment]
I think you can, yes. That's what @asmeurer found out and what allowed him to create this mechanism.

I'm not completely sure about the details but I remember when @asmeurer mentioned this bit of information :-)

**Comment by msarahan** [comment]
yes, you're right.  I have tested by running Python, and moving the DLLs folder and all DLLs in the root folder to a separate folder.

Still, these sporadic issues with permissions always point to some DLL that's in use.  For example, the Anaconda Build team has had a terrible time on Windows with this exact kind of error.  I think there's more to this story.  I don't know when an application "locks a DLL": http://stackoverflow.com/questions/4074480/can-i-safely-rename-move-a-dll-that-is-already-loaded-by-a-process - it does not appear to lock DLLs simply by loading them.

**Comment by ccordoba12** [comment]
But we haven't heard (at least I haven't :-) from other issues like the one reported here by @ukoethe. And this moving-files-in-use mechanism has been in place for more than six months.

So I still think it has to do with a system wide installation and conda not having the right permissions to move files.

**Comment by msarahan** [comment]
> But we haven't heard (at least I haven't :-) from other issues like the one reported here by @ukoethe.

That's not quite accurate:

https://github.com/conda/conda/issues/2003
https://github.com/conda/conda/issues/1871

There are others that mention other files in use, but those two issues specifically mention the DLLs folder.

For the record, Aaron's PR was: https://github.com/conda/conda/pull/1133 - merged in May 2015

And here's a Stack Overflow post about it: http://serverfault.com/questions/503721/replacing-dll-files-while-the-application-is-running - what I took from it is "YMMV"

**Comment by ccordoba12** [comment]
> That's not quite accurate

Ok, didn't know that.

> For the record, Aaron's PR was: #1133 - merged in May 2015

Yes, but it wasn't used to replace Python (and maybe other packages in the root env?) until some months later :-)

----

But in the end it seems you're right: there's something fishy going on when trying to move DLLs. Really hard problem to tackle...

**Comment by msarahan** [comment]
I think this isn't too bad - we probably just need to wrap the move attempt in a try-catch, with the catch falling back to moving the file on reboot.  If any files are moved on reboot, advise the user that a reboot is needed.  Might need to also leave a temporary file around to indicate to new Python instances that a reboot is needed, as things may or may not work properly until the reboot completes.

**Comment by ukoethe** [comment]
> So I still think it has to do with a system wide installation and conda not having the right permissions to move files.

But then every attempt with insufficient privileges should fail, shouldn't it? I retried with a fresh system-wide installation, and the called `conda update conda` with normal user privileges, and it worked. So I tend to believe @msarahan's race condition theory.

**Comment by ukoethe** [comment]
Requiring a reboot for something like `conda update conda` is an **absolute no-go**! People accept that OS updates may require a reboot, but everything else should be able to do it without. Acrobat Reader is about the only program that's too stuoid to do it without a reboot, and everyone hates this.

BTW, I think what failed was the Python update, not the conda update (but I'm not sure about this). 

A simple and safe solution to update Python in the root environment would be a little script `update-conda.bat` that creates a temporary Python environment containing the new Python, and then run the update of the root environment from there. Calling `conda update python` or similar in the root environment would then simply issue the error message "Please use 'update-conda'  to update the root environment."

**Comment by ukoethe** [comment]
Can you also offer advice how to recover from the update failure, short of starting from scratch? In the case I reported, the problem fortunately occured in a test installation that I could easily recreate. But had it happened in my main installation (with several carefully constructued environments for production use), it would have been a real setback.

**Comment by msarahan** [comment]
Yes, that's a good idea.  Another (cleaner, IMHO) idea would be to separate Conda completely from the root environment, as proposed in https://github.com/conda/conda/wiki/CEP-1:-Replacing-root-environment-with-default-environment and https://github.com/conda/conda/issues/1022

I don't think it matters much how it is implemented (either as a frozen EXE, or as a separate Python install), but it definitely can't share files with any environment that it is managing.  I think it would get updated in exactly the manner you propose: a separate helper script, or dropping in a new frozen EXE.

I think there's been more recent discussion on this, but I can't seem to find it.  I think @groutr proposed it.

@kalefranz please weigh in.

**Comment by msarahan** [comment]
@ukoethe for reconstructing a broken install:

1. All of your environments are OK.  They all live in $PREFIX/envs (unless you otherwise specify).  You can copy those folders into anywhere else and they should be OK.
2. There is a "recovery mode" in the installer scripts that clobbers the existing install, overwriting whatever the installer contains.  It should make things work again, but leaves the root env in a strange, incomplete metadata state.  Pass the -f flag (for force) to the shell installers.  I'm not sure if a Windows equivalent exists right now.

**Comment by insertinterestingnamehere** [comment]
My understanding of the linking issue was that filesystem links are used for nearly everything and that, when a dll is in use, the link originally followed to get to it can still be updated. As long as nothing is touching the actual location of the dll, stuff like this shouldn't happen.
I'm really curious why the error points to the whole dlls directory rather than any dll in particular though. That has never been the case in the errors like this that we've worked through previously.

**Comment by groutr** [comment]
FWIW, I ran into this issue the other week a couple of times on my windows build machine.  However, I wasn't able to reliable reproduce it or trace it down to anything.

**Comment by ijstokes** [comment]
I believe this is another argument for the discussion centered around #1329 (conda environments accessed by multiple users)

@msarahan and @groutr more discussion on this in #1733 

@ukoethe I added a comment in #1733 back in October that might give you some advice on how to recover from this situation.  I've been in it many times, and it is always somewhat painful.  I was happy when I came up with that solution, though there may be some issues translating it to a Windows environment.  Maybe msysgit or cygwin will give you the tools to do it (mostly it comes down to the rsync part).

**Comment by rekcahpassyla** [comment]
This has also now just happened to me for the second time. I have administrator rights on my machine.; I've never had to open an administrator terminal to conda update or install before. This is the second time I've had to reinstall anaconda - I now get this

```python

C:\dev\code>conda
ImportError: No module named site


C:\dev\code>python
'python' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by shahlaebrahimi** [comment]
I would be very grateful if you could let me know how to solve this issue. In fact, I tried the following command in order to install scikit 0.18. 

` conda install -c anaconda scikit-learn=0.18`

After installation, I get 

```
C:\Users\Markazi.co\Anaconda3\python.exe D:/mifs-master_2/MU/NeuralNetwork.py
Traceback (most recent call last):
  File "D:/mifs-master_2/MU/NeuralNetwork.py", line 14, in <module>
    from sklearn.model_selection import cross_val_score, StratifiedKFold,train_test_split
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\sklearn\__init__.py", line 57, in <module>
    from .base import clone
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\sklearn\base.py", line 12, in <module>
    from .utils.fixes import signature
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\sklearn\utils\__init__.py", line 11, in <module>
    from .validation import (as_float_array,
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\sklearn\utils\validation.py", line 18, in <module>
    from ..utils.fixes import signature
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\sklearn\utils\fixes.py", line 290, in <module>
    from scipy.sparse.linalg import lsqr as sparse_lsqr
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\scipy\sparse\linalg\__init__.py", line 112, in <module>
    from .isolve import *
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\scipy\sparse\linalg\isolve\__init__.py", line 6, in <module>
    from .iterative import *
  File "C:\Users\Markazi.co\Anaconda3\lib\site-packages\scipy\sparse\linalg\isolve\iterative.py", line 7, in <module>
    from . import _iterative
ImportError: DLL load failed: The specified module could not be found.

Process finished with exit code 1

```

Thanks in advance.
Best regards,

**Comment by kalefranz** [comment]
fixed with conda 4.3
